backend/common/utils/spatial.py

```diff
@@ -0,0 +1,158 @@
+import io
+import logging
+import os
+import shutil
+
+import numpy as np
+import pyvips
+from PIL import Image
+
+from backend.layers.thirdparty.s3_provider import S3Provider
+
+logger = logging.getLogger(__name__)
+
+
+class SpatialDataProcessor:
+    """
+    A class that processes spatial data and generates deep zoom assets.
+
+    Args:
+        s3_provider (S3Provider, optional): An instance of the S3Provider class. Defaults to None.
+
+    Attributes:
+        s3_provider (S3Provider): An instance of the S3Provider class.
+        deployment_stage (str): The deployment stage (e.g., "staging").
+        env (str): The environment based on the deployment stage.
+        bucket_name (str): The name of the S3 bucket.
+        asset_directory (str): The directory for the deep zoom assets.
+    """
+
+    def __init__(self, s3_provider: S3Provider = None):
+        self.s3_provider = s3_provider if s3_provider else S3Provider()
+        self.deployment_stage = os.getenv("DEPLOYMENT_STAGE", "staging")
+        # relying on dev bucket for rdev and test
+        self.env = "dev" if self.deployment_stage in ["rdev", "test"] else self.deployment_stage
+        self.bucket_name = f"spatial-deepzoom-{self.env}"
+        self.asset_directory = "spatial-deep-zoom"
+
+    def _calculate_aspect_ratio_crop(self, image_size):
+        """
+        Calculate the aspect ratio crop coordinates for an image.
+
+        Args:
+            image_size (int, int): The size of the image (width, height).
+
+        Returns:
+            tuple: The crop coordinates (left, upper, right, lower).
+        """
+        width, height = image_size
+        new_dimension = min(width, height)
+        left = (width - new_dimension) / 2
+        upper = (height - new_dimension) / 2
+        right = (width + new_dimension) / 2
+        lower = (height + new_dimension) / 2
+        return tuple(map(int, (left, upper, right, lower)))
+
+    def _prepare_image(self, content):
+        """
+        Prepare the image array for processing.
+
+        Args:
+            content (dict): The content dictionary containing the image array.
+
+        Returns:
+            np.ndarray: The prepared image array.
+        """
+        resolution = "fullres" if "fullres" in content["images"] else "hires"
+        image_array = content["images"][resolution]
+        image_array_uint8 = np.uint8(image_array * 255 if resolution == "hires" else image_array)
```

Comment:
`hires` image array pixel intensity values are commonly normalized by floating point value range [0, 1] 
`fullres` values range [0, 255] - 0 representing black and 255 white or max intensity of a color channel
- since we are passing uint8 (8-bit unsigned integer) array to PIL, this line upscales `hires` data to full range between 0 and 255, to ensure that `hires` array fits the data type of uint8
'fullres` is already in the desired range so there's no need to upscale.

for example, if a `hires` image array has a value of 0.5:

0.5 * 255 = 127.5

Converting to uint8: np.uint8(127.5) = 127

Ensuring that values are in the range [0, 255] for uint8.

Reply:
@kaloster - both [fullres](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/5.1.0/schema.md#spatiallibrary_idimagesfullres) and [hires](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/5.1.0/schema.md#spatiallibrary_idimageshires) are _normalized_  per:
`The numpy.ndarray.dtype MUST be numpy.uint8
`

Reply:
Thanks, @brianraymor ! Does that mean both hires and fullres need `* 255`?

Reply:
> Does that mean both hires and fullres need * 255?

No, it means that neither need `* 255` because both must be unsigned ints rather than floats.

Reply:
Ok thanks - this was based on the current data I was testing this with - will remove the `_prepare_image()` method